Readme.md

```diff
@@ -0,0 +1,1432 @@
+
+`Option`s returned by functions force you to check for the absence of value before using it. This can be done using pattern matching:
+
+```scala
+val v: Vector[Int] = ...
+v.find(_ > 0) match {
+  case Some(n) => println("$n is positive")
+  case None => println("No positive values")
+}
+```
+
+Moreover, `Option` has several methods which make working with possibly absent values easier:
+
+```scala
+// when we only need to handle the "value is present" case
+v.find(_ > 0).foreach { n =>
+  println("$n is positive")
+}
+
+// when we need to return a default value in case
+// the computed value is absent
+val n = v.find(_ > 0).getOrElse(0)
+```
+
+When interfacing with Java code, all values coming from the Java side must be checked for `null`s if it is not stated explicitly that no `null` values are possible. `Option.apply` methods accepts a nullable value and returns an `Option` which will be `None` if the passed value is actually `null`:
+
+```scala
+Option(javaClass.getResourceAsStream("some_file"))
+```
+
+Do not call `get` on optional values except when you're absolutely sure that the optional value is always non-empty. Often such code may be rewritten more clearly using pattern matching or other optional combinators.
+
+### Recursion
+
+Recursion is often very natural for certain algorithms, especially on recursive structures like trees and lists. In many cases, however, recursion tends to make the code obscure.
+
+Always try to solve the problem without explicit recursion first. Scala collections library provides lots of combinators which make explicit recursive traversals unnecessary in many cases.
+
+If a problem cannot be solved with pre-defined, combinators, compare the imperative version using loops and recursion first. In many cases, especially when the logic is not very large, imperative implementation may be simpler than recursion. In other cases, recursion would be clearer.
+
+When writing recursive functions, try to write them in a tail-recursive way, enforced by the `@tailrec` annotation. Tail-recursive functions will be optimized by the compiler into an imperative loop. However, tail-recursive functions may require passing lots of state in an accumulator parameter. Sometimes rewriting such functions using imperative loops may make the code nicer.
+
+## Object-oriented programming
+
+Scala is a powerful object-oriented language; if fact, it is even more pure than Java. Everything in Scala is an object, and its facilities for object-oriented design are very expressive and convenient. Naturally, most of object-oriented design practices are possible in Scala; they are described in numerous forms in various books and articles and therefore are omitted here. In this document we will examine Scala-specific features which help with the object-oriented programming.
+
+### Dependency injection
```

Comment:
After long though I still cannot formulate something along these lines which I would agree with.

First of all, I don't think it is bad to use manual DI or even a DI framework to construct an object graph for an entire process if this process is small enough, and in microservices/service-oriented architectures processes do tend to be relatively small. So I would agree with all these statements if they were prepended with a phrase like "In a monolithic architecture, ...".

But even in case with monolithic architecture, I wouldn't really agree with the necessity to instantiate subgraphs for each module separately. In a complex application the graph of modules and therefore the object graph tends to be quite large, with different parts having different lifecycles. Modules are also interconnected with each other, which requires injecting modules as dependencies into each other. I don't think I see the sufficient justification to do it manually.

Moreover, many DI frameworks, which in case of monolithic architecture would be anyway necessary, already provide means for isolation of modules. For example, Guice and Guice-like frameworks has explicit modules, some of which could be "private" (they export only those dependencies which are explicitly declared as exported), and sub-injectors for further isolation of module subgraphs. I believe Spring has something like it in the form of nested contexts.

The only problem I can see with DI and testing is uncontrolled growth of dependencies of certain classes. If a class has lots of dependencies and its implementation tend to use all these dependencies at once, it could be quite hard to test it. But this is not really related to DI itself and the level of isolation of modules and how they are created - it is a consequence of violation of a single responsibility principle, where growing responsibilities of a class require more dependencies. This problem is somewhat orthogonal to the DI, it is just a consequence of badly distributed consequences and would happen regardless of the level on which the DI is used.

It is also possible that I misunderstood what you meant, so please do tell if I did get something wrong :)